Visitor can read one article

rithreaksa opened the pull request:
[PT-story](https://www.pivotaltracker.com/story/show/176031727)
### User Story
```
As a Visitor
In order to read an article
I would like to be able to click on an article and have it displayed

```
## Changes proposed in this pull request:

* Adds visitor can read specific article feature
* Adds 'Read More' and 'back' buttons
* Refactors to use redux

## What I have learned working on this feature:
* Use state with redux
* Importance of name convention

## Packages/Gems added
* Adds redux

## Screenshots (Client)

<img width="1280" alt="Screen Shot 2021-01-07 at 11 39 19 AM" src="https://user-images.githubusercontent.com/62254746/103883096-082b9300-50dd-11eb-86a9-f78f227aaa24.png">
<img width="1280" alt="Screen Shot 2021-01-07 at 11 39 04 AM" src="https://user-images.githubusercontent.com/62254746/103883112-0e217400-50dd-11eb-8118-18ff005f1501.png">

ViktorHek committed:
creating fixture with one article. creating test to read one article. creating ArticleCard. adding Read More button in DisplayArticle

ViktorHek committed:
comenting out ArticleCard bc we do not need it atm. adding fetchSingleArticle function in DisplayArticlesData. createt GET request with ID-param in articleData

KyunginNa committed:
Modifying key name in article_data.json
Modifying fetchSingleArticle function
Modifying visitorCanReadOneArticle.feature.js

KyunginNa committed:
Modifying fixture file
Modifying visitorCanReadOneArticle.feature.js
Deleting map for articleDetails and adding it in return block

ViktorHek committed:
adding go-back button

ViktorHek committed:
adding redux. creating indexCrad and singleArticleCard

ViktorHek committed:
readux setup

ViktorHek committed:
slight modification visitorCanViewAllArticles.
adding props to articleCard and change nameing conventions
refactoring articleCard using redux
comment out singleArticle.jsadding rootReducer
adding initialState

ViktorHek committed:
cleaning up a little

rithreaksa committed:
Refactors DisplayArticle, SingleArticle and ArticleCard components

rithreaksa committed:
Modifies state in DisplayArticle component
Fix styling in SingleArticle component

rithreaksa committed:
Adds styling to SingleArticle component

emtalen commented on src/components/DisplayArticlesData.jsx:
```diff
@@ -1,60 +1,43 @@
-    setArticlesData(result.data.articles);
+    const result = await getArticleData();
+    dispatch({ type: 'SET_ARTICLES_INDEX', payload: result.data.articles });
+    dispatch({ type: 'SET_SINGLE_ARTICLE' });
```
why are you dispatching for a single article here?
I would either pass dispatch as props to the module here, so you can dispatch it directly that way, or do as Thomas showed you yesterday to refactor the store so you can access it from a module. 
In that case you would make the `getArticleData` global to use from anywhere and you don't have to pass that function around but can simply import it where you need it

rithreaksa replied:
We refactored that. Thanks.

emtalen commented on src/components/DisplayArticlesData.jsx:
```diff
@@ -1,60 +1,43 @@
+  const fetchSingleArticle = async (event) => {
+    const id = event.target.dataset.id;
+    const response = await getSpecificArticle(id);
+    dispatch({ type: 'SET_SINGLE_ARTICLE', payload: response.data.article });
```
If you are setting the response from this function in application state, you don't need to pass it as props but you can execute it on the level where it is being triggered, in the article card

rithreaksa replied:
We fixed this one.

emtalen commented on src/components/DisplayArticlesData.jsx:
```diff
@@ -1,60 +1,43 @@
+          />
+        </Card.Group>
+      )}
+      <SingleArticle singleArticle={singleArticle} getArticles={getArticles} />
```
You can subscribe to the state of a single article in the component instead of passing it as props

rithreaksa replied:
We also refactored this one too.

emtalen commented on src/components/SingleArticle.jsx:
```diff
@@ -0,0 +1,30 @@
+const SingleArticle = ({ singleArticle, getArticles }) => {
+  return (
+    <>
+      {singleArticle && (
```
put the singleArticle as an empty object in your initial state, then you can remove this conditional

rithreaksa replied:
Yep we did that now. Thank you.

sealfury committed:
refactoring module to include dispatch functionality
first steps refactoring singleArticle module

sealfury committed:
configuring browser router and adding useParams hook, adding singleArticle to initial state

sealfury committed:
moving functionality from article card to article display
adjusting naming conventions for clarity
adding navlink routing to article card

FlisAnn committed:
Adding NavLink to the back-button

FlisAnn committed:
Adding test for not finding article.
All test are running green.

emtalen commented on cypress/integration/visitorCanReadOneArticle.feature.js:
```diff
@@ -0,0 +1,55 @@
+      cy.get("[data-cy='read-me-button2']").click();
+    });
+    it("and is presented with an error message", () => {
+      cy.get("[data-cy='error-message-article']").should("contain", "Article");
```
please look that it contains the entire error message

emtalen commented on cypress/integration/visitorCanReadOneArticle.feature.js:
```diff
@@ -0,0 +1,55 @@
+      cy.server();
+      cy.route({
+        method: "GET",
+        url: "http://localhost:3000/api/articles/2",
```
why are you specifying the url here with the ID but use a wild card for the happy path?

sealfury replied:
using wild card for both now

emtalen commented on src/components/DisplayArticlesData.jsx:
```diff
@@ -1,60 +1,28 @@
 
   return (
-    <>{articlesData.length > 0 && <div data-cy="index">{articleIndex}</div>}</>
+    <>{mainArticles.length > 0 && <ul data-cy="index">{articleIndex}</ul>}</>
```
why are you using an `ul` when the `articleIndex` does not contain list items but cards?

sealfury replied:
changed it to a `div`

emtalen commented on src/index.js:
```diff
@@ -1,17 +1,29 @@
+    <BrowserRouter> 
+      <App />
+    </BrowserRouter>
+   
```
remove xtra line please

emtalen commented on src/state/store/configureStore.js:
```diff
@@ -0,0 +1,7 @@
+import rootReducer from "../reducers/rootReducer";
+import initialState from "./initialState"
+
+const store = createStore(rootReducer, initialState);
```
when you are creating the store like this you don't need to pass the `useDispatch` hook as an argument but in your module you can simply import the store and do `store.dispatch({type: blabkla, payload: blabla})`

sealfury committed:
cleaning and formatting code editing for consistency in test stubs
replacing dispatch hooks/arguments with importing store to module
